ce/api/streamflow/watershed_streams.py

```diff
@@ -0,0 +1,164 @@
+from ce.geo_data_grid_2d.vic import VicDataGrid
+from ce.api.streamflow.shared import (
+    setup,
+    is_upstream,
+    VIC_direction_matrix,
+    get_time_invariant_variable_dataset,
+)
+
+
+def watershed_streams(sesh, station, ensemble_name):
+    """Return a GeoJSON multi line string object representing the different streams
+    in a watershed that are upstream of the inputted point.
+
+    :param sesh: (sqlalchemy.orm.session.Session) A database Session object
+    :param station: (string) Location of drainage point, WKT POINT format
+    :param ensemble_name: (string) Name of the ensemble containing data files backing
+        providing data for this request.
+    :return: dict representation for JSON response object with the following
+        attributes:
+            Lines: A GeoJSON MultiLineString representing the streams of the watershed;
+
+    This function is primarily responsible for finding the relevant data files
+    and converting their contents to `VicDataGrid` objects for consumption by
+    `worker`, which as its name suggests, does most of the work.
+    """
+    station_lonlat = setup(station)
+
+    with get_time_invariant_variable_dataset(
+        sesh, ensemble_name, "flow_direction"
+    ) as flow_direction_ds:
+        try:
+            return worker(
+                station_lonlat,
+                flow_direction=VicDataGrid.from_nc_dataset(
+                    flow_direction_ds, "flow_direction"
+                ),
+            )
+        except GeoDataGrid2DIndexError:
+            abort(
+                404,
+                description="Station lon-lat coordinates are not within the area "
+                "for which we have data.",
+            )
+
+
+def worker(
+    station_lonlat, flow_direction,
+):
+    """Returns the same as watershed_streams, this function exists to make 
+    these computations more easily testable. (Specifically, data *files* are not 
+    required, only the relevant contents of those files passed as `VicDataGrid` objects.
+    `VicDataGrid`s are much easier to construct for tests.)
+
+    :param station_lonlat: (tuple) Location of drainage point, (lon, lat)
+    :param flow_direction: (VicDataGrid) Flow direction grid
+    :return: (dict) representation for JSON response object; see watershed() for details
+    """
+
+    # Compute lonlats of watershed whose mouth is at `station`
+    # TODO: Refactor to accept a VicDataGrid?
+    direction_matrix = VIC_direction_matrix(
+        flow_direction.lat_step, flow_direction.lon_step
+    )
+
+    with Timer() as watershed_time:
+        watershed_xys = build_watershed_streams(
+            flow_direction.lonlat_to_xy(station_lonlat),
+            flow_direction.values,
+            direction_matrix,
+            debug=True,
+        )
+
+    # `watershed_lonlats` must be an ordered collection (not sets) because
+    # a multi line string is an array (python list) of linestrings
+    watershed_lonlats = [[]]
```

Comment:
We no longer need this line; it's a leftover from the previous way of computing `watershed_lonlats` .